pkg/promotion/releasepromotion/mirror.go

```diff
@@ -0,0 +1,193 @@
+	SkipNames []string `json:"skip_names"`
+}
+
+func (t mirrorConfigTarget) IsSkipped(name string) bool {
+	for _, skip := range t.SkipNames {
+		if strings.Contains(name, skip) {
+			return true
+		}
+	}
+	return false
+}
+
+func readMirrorConfig(name string) (*mirrorConfig, error) {
+	data, err := ioutil.ReadFile(name)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	var config mirrorConfig
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+func loadMirrorConfig(baseDir string, parts []string) (*mirrorConfig, error) {
+	config, err := readMirrorConfig(filepath.Join(append(append([]string{baseDir}, parts...), "mirroring.yaml")...))
+	if err != nil {
+		return nil, err
+	}
+	if config != nil {
+		return config, nil
+	}
+	if len(parts) == 0 {
+		return nil, nil
+	}
+	return loadMirrorConfig(baseDir, parts[:len(parts)-1])
+}
+
+// WriteMirroringRules uses the mirroring.yaml files within a directory hierarchy to
+// generate the mirroring rules for that release from a given source cluster to a target
+// external registry.
+func (m *Verifier) WriteMirroringRules(toDir, commandName string) error {
+	// remove any old names under this directory
+	if err := filepath.Walk(toDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if name := filepath.Base(info.Name()); name == "mirroring.rules" {
+			return os.Remove(path)
+		}
+		return nil
+	}); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	// write mirroring rules
+	for key, target := range m.targets {
+		if len(target) == 0 {
+			continue
+		}
+		// find the mirroring config from the current dir up to the release root
+		config, err := loadMirrorConfig(toDir, key.Directory())
+		if err != nil {
+			return err
+		}
+		if config == nil {
+			continue
+		}
+
+		dir := filepath.Join(append([]string{toDir}, key.Directory()...)...)
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			return err
+		}
+
+		names := sets.NewString()
+		for key := range target {
+			names.Insert(key)
+		}
+		streamName := key.Name
+		if len(streamName) == 0 {
+			streamName = key.Tag
+		}
+		sourceLocation := key.PullSpec()
+
+		buf := &bytes.Buffer{}
+		fmt.Fprintf(buf, "# generated by %s, do not edit\n", commandName)
+
+		for _, name := range names.List() {
+			for _, target := range config.Targets {
+				if target.IsSkipped(name) {
+					continue
+				}
+				versions := target.Versions
+				if target.VersionLatest == streamName {
+					versions = append(versions, "latest")
+				}
+
+				fmt.Fprintf(buf, "%s", strings.NewReplacer("{stream}", streamName, "{image}", name).Replace(sourceLocation))
+				for _, version := range versions {
+					for _, pattern := range target.Patterns {
+						r := strings.NewReplacer("{version}", version, "{stream}", streamName, "{image}", name)
```

Comment:
nit: either name the replacer in all cases or don't, for readability and parallelism